lib/signs/utilities/messages.ex

```diff
@@ -258,6 +304,73 @@ defmodule Signs.Utilities.Messages do
     end
   end
 
+  @spec overnight_period?(
+          DateTime.t(),
+          DateTime.t(),
+          DateTime.t(),
+          DateTime.t(),
+          boolean()
+        ) :: boolean()
+  defp overnight_period?(
+         _current_time,
+         first_scheduled_departure,
+         last_scheduled_departure,
+         _most_recent_departure,
+         false
+       )
+       when first_scheduled_departure == nil or last_scheduled_departure == nil,
+       do: false
+
+  defp overnight_period?(
+         current_time,
+         first_scheduled_departure,
+         last_scheduled_departure,
+         _most_recent_departure,
+         false
+       ),
+       do:
+         calculate_overnight_period(
+           last_scheduled_departure,
+           current_time,
+           first_scheduled_departure
+         )
+
+  defp overnight_period?(
+         current_time,
+         first_scheduled_departure,
+         last_scheduled_departure,
+         most_recent_departure,
+         true
+       ) do
+    if Timex.after?(most_recent_departure, last_scheduled_departure) do
+      most_recent_departure
```

Comment:
I'd assume that if we got a definitive last-trip tag, we'd want to use that time as the end of service, even if it's before the scheduled end, right?